sys/kern/src/arch/rv64/stimer.rs

```diff
@@ -2,25 +2,31 @@
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
+use crate::arch::clock_freq::CLOCK_FREQ_KHZ;
 use crate::arch::sbi_set_timer;
 
-const HAS_SSTC: bool = false; // TODO: Runtime detection
-
 #[no_mangle]
-pub unsafe fn set_timer(tick_divisor: u32) {
-    // TODO: Feature detection for Sstc, SBI call for non-supporting, supporting get:
-    //       riscv::register::stimecmp::write(tick_divisor.into());
-    if HAS_SSTC {
-        riscv::register::stimecmp::write(tick_divisor as usize)
+pub unsafe fn set_timer() {
+    let current = riscv::register::time::read();
+
+    if cfg!(feature = "riscv-support-sstc") {
+        riscv::register::stimecmp::write(current)
```

Comment:
I think stimecmp should be set to current + tick_divisor. Else, interrupt will fire right away. Same for sbi_set_timer below.

Reply:
This behaviour is, I believe, largely identically to the RV32 M-Mode:

`        // make MTIMECMP start with MTIME
        let mtime = core::ptr::read_volatile(crate::startup::MTIME as *mut u64);
        core::ptr::write_volatile(crate::startup::MTIMECMP as *mut u64, mtime);

        // increment mtimecmp for appropriate timer interrupts
        reset_timer();`

Further - the only call to set_timer is from task.rs::start_first_task, which only enables the interrupt after setting the initial value of xtimecmp.

Reply:
The difference between RV32 and RV64 is:

On RV32, mtimecmp is set to mtime and then immediately it calls `reset_timer()` before enabling timer interrupts.

On RV64, `set_timer()` call isn't followed by `reset_timer()`. Instead it enables timer interrupts directly. This will result in a pre-mature interrupt.